test/pylib/manager_client.py

```diff
@@ -255,8 +255,9 @@ async def host_is_known():
 
     async def wait_for_host_down(self, dst_server_id: str, server_id: str, deadline: Optional[float] = None) -> None:
```

Comment:
This function is broken, its callsite is broken, and everything is broken. (Not your fault, it was broken before.)
@kostja @gusev-p @alecco just look at this beautiful world of untyped Python.

The only callsite of the function is in test/topology/test_topology_remove_decom.py:
```python3
            await manager.wait_for_host_down(initiator_ip, target_ip)
```

the variables passed have "ip" in their names.

If we assume that the function is operating on IP addresses, just that the parameters names are broken, then its implementation is correct already. In particular, `get_down_endpoints` returns a list of IP addresses, so `server_id in down_endpoints` check is correct assuming that `server_id` is actually an IP address.

So I immediately though that we should leave the logic of this function unchanged, just change the names of parameters to have `ip` instead of `id` in their names.

However, if we dig deeper, things get even more interesting :)

Because `initiator_ip` and `target_ip` at the callsite are not actually IP addresses!
```python3
            server_ids = await manager.running_servers()
            host_ids = await asyncio.gather(*(manager.get_host_id(s) for s in server_ids))
            initiator_index, target_index = random.sample(range(len(server_ids)), 2)
            initiator_ip = server_ids[initiator_index]
            target_ip = server_ids[target_index]
```

The name `server_ids` suggests that `server_ids` is a list of IDs.
The line `initiator_ip = server_ids[initiator_index]` suggests that it's actually a list of IPs.
But what does `await manager.running_servers()` actually return?
```python3
    async def running_servers(self) -> List[ServerInfo]:
```
a list of `ServerInfo`s.

Beautiful.

Every single thing here is damn broken. Nothing matches anything!

Of course the test `test_remove_node_with_concurrent_ddl` is completely broken, we never run it.

Also note that `get_down_endpoints` is **already** operating on IP addresses even before this commit, because `self.client.get` accepts only an IP address. It's just that the name of the parameter is broken.

@Deexie please modify this commit to do a larger cleanup:
- modify names of parameters in `get_down_endpoints` and `wait_for_host_down` to mention **IP**s, not IDs
- instead of using `str` for the type hints, use `IPAddress` from `test.pylib.internal_types` module
- fix `test_remove_node_with_concurrent_ddl` in test/topology/test_topology_remove_decom.py

for fixing `test_remove_node_with_concurrent_ddl` I suggest the following:
- add return type hints to `do_ddl()` and `do_remove_node()` like this:
```python3
    async def do_remove_node() -> None:
```
- this will allow `mypy` to catch all type errors in this function. Run it on the file.
- fix the type errors and variable names correspondingly.

Reply:
Python is evil. I would rather write tests in C++, which is actually what I did at my previous work place, and it was way more effective that this. I don't know why people think tests should only be written in Python.

Another problem is that the `test_remove_node_with_concurrent_ddl` is skipped, we should unskip it. I wrote this test some time ago, then got distracted on that `@slow` attribute thing, and then drowned in Jenkins configs in `scylla-pkg` repo, so the PR is not merged and this attribute still doesn't work on Jenkins. I think we can still remove `@pytest.mark.skip` attribute, just reduce the amount of iterations in `do_remove_node` so that the test doesn't take ages.

@Deexie The test `test_remove_node_with_concurrent_ddl` used to pass when I wrote it, but if there would be any problems with it, feel free to contact me)

Reply:
@gusev-p I'm really surprised by your claim that writing tests in C++ is more effective than writing tests in Python, ad that Python is "evil". I want to explain why I think you're wrong, and what might have led you to this wrong (in my opinion) conclusion:

Clearly, if you want to write unit tests for some internal C++ functions - e.g., testing the internal implementation of the semaphore or loading cache - then writing the test in C++ is your best, or only option. But if you're writing functional tests - tests which aim to check **user-visible** features, those can be written in **any** language because they interact with Scylla through the user-visible protocol - not through C++. In my exprience, C++ is not a great language to write the functional tests in. We actually have a lot of experience writing such tests in C++ in Scylla - we have in test/boost/ many functional tests in C++ that use what we called "cql_env" to write functional tests in C++ that interact with CQL.  I would estimate that we have around 40,000 lines of functional test code using this cql_env approach, and you are welcome to write more if that's what you like. But these C++ tests have some pretty big drawbacks that I can tell you from personal experience:

1. The tests need slow compilation which make them annoying to write and make people less likely to run other people's tests manually.
2. When a C++ test fails it's often hard to tell why. The Boost test framework assert feature is crappy compared to pytest, and when a failure results in a crash dump it's even less obvious which point the test reached, because the test crashed together with the tested system.
3. You often need a lot of boilerplate code to write a simple test. E.g., to compare the result of a query to expected results, you need to write these expected results using ugly and verbose C++ functions.
4. Moreover, you're never sure if you actually used those C++ functions correctly and if you see a bug, if it's a bug in Scylla or the test. Even worse, there are cases like https://github.com/scylladb/scylladb/issues/4126 where the C++ functions don't do exactly what ordinary users would see, so your test doesn't even check the real behavior of Scylla.
5. When you write a test in C++ which calls internal Scylla functions, you cannot run this test against Cassandra or old versions of Scylla. Many times in the past we had tests that actually enshrined wrong behavior, which we would have known had we run the same test against Cassandra.

In contrast, my experience in writing tests in Python, is the complete opposite. Tests are easy and almost "fun" to write, fast to write, easy to debug, test real user-visible features (and not internal details of the test framework), and easy to run against competing implementations (like Cassandra or DynamoDB) or older versions of Scylla.
 
Remember that above I mentioned that in the last 8 years all Scylla developers together wrote about 40,000 of lines of C++ test code for CQL functional tests?  Well, in the last three years, I single-handedly wrote **more** than 40,000 lines of Python test code, in test/alternator and test/cql-pytest. It is just so much easier, so much more fun, so much more effective and so much more productive to write these tests, that I just couldn't stop :-)

That being said, it's not enough to just use Python and not use C++. You also need to be aware of how you use it and why. dtest is a notoriously (in my opinion) bad example of tests in Python. For many years (until a few weeks ago, basically), running any test in dtest took as much as 5 minutes (!) to start, making it even slower than compiling a C++ test. dtest uses a huge library of "convenience functions" which nobody fully understands and each does too many things and does them badly. Dtest doesn't (until recently) use fixture so it encourages very long tests that test many different things in one test and are very hard to understand when something fails. I took all those things as examples of what NOT to do in test/cql-pytest when I designed it - please take a look at last section of test/cql-pytest/README.md ("Developing new cql-pytest tests") for cql-pytest's philosophy. To make a long story short, cql-pytest has a small number of convenience functions, and most tests just use the CQL driver in a way any user application would use it. cql-pytest encourages many small tests that use fixture. Running a single cql-pytest test usually takes less than a second (!).

Note that I didn't mention above any word about "strong typing" solving anything. You and Kamil seem to think this has huge importance - and I think it does not. At all. Yes, part of the "experience" of writing a new test in Python is that it doesn't work because of stupid typing errors, and yes, some of these stupid errors are made possible because of Python's weak typing. So what? Since it's so easy to run and re-run your test, and since pytest assertions and Python failures are usually very easy to debug (containing very useful messages and backtrace), you usually find these errors in a minute. Yes, mypy could have helped you find these errors before you even ran the test once, but... so what? At the same time, using mypy makes all the test code harder to write, harder to read. I honestly see it as a losing proposition. 

You can argue that mypy is good for the "library" part of the test - the parts that many tests need to share so it's better that we write it well. Here I bring you back to my philosophy that the amount of this "library" stuff should be minimized.  For example, in test/alternator we have 18,000 lines of code of actual tests (test/alternator/test*.py) and just 1172 lines of code for the library and framework (test/cql-pytest/run.py, test/alternator/run, test/alternator/contest.py, test/alternator/util.py and test/alternator/scylla_log.py). That's a 6% library-to-actual-tests ratio. Contrast this with what you guys have been doing on topology tests - 5000 lines of code for the framework (test.py and test/pylib/**/*.py) and just 3300 lines of code for the tests (test/topology*, test/object_store, test/broadcast_tables - together). That's a 150% library-to-actual-tests ratio! I think that **this** is the reason why you're not enjoying writing these tests. It's not the Python language. You're simply writing to few tests to realize how easy it can be to write them.

Reply:
> You can argue that mypy is good for the "library" part of the test - the parts that many tests need to share so it's better that we write it well. Here I bring you back to my philosophy that the amount of this "library" stuff should be minimized. For example, in test/alternator we have 18,000 lines of code of actual tests (test/alternator/test*.py) and just 1172 lines of code for the library and framework (test/cql-pytest/run.py, test/alternator/run, test/alternator/contest.py, test/alternator/util.py and test/alternator/scylla_log.py). That's a 6% library-to-actual-tests ratio. Contrast this with what you guys have been doing on topology tests - 5000 lines of code for the framework (test.py and test/pylib/**/.py) and just 3300 lines of code for the tests (test/topology, test/object_store, test/broadcast_tables - together). That's a 150% library-to-actual-tests ratio! I think that this is the reason why you're not enjoying writing these tests. It's not the Python language. You're simply writing to few tests to realize how easy it can be to write them.

Maybe we should just stop doing cluster tests altogether because they are not fun and require a lot of scaffolding to manage clusters. Indeed writing simple single-node tests is much more fun, so we should do only that. After all, single-node Scylla deployments are the most common ones out there.

Reply:
> Yes, part of the "experience" of writing a new test in Python is that it doesn't work because of stupid typing errors, and yes, some of these stupid errors are made possible because of Python's weak typing. So what? Since it's so easy to run and re-run your test, and since pytest assertions and Python failures are usually very easy to debug (containing very useful messages and backtrace), you usually find these errors in a minute.

Rarely I read thing so much detached from reality :) Sorry I could not resist. In my limited experience with the Python (which I hate with a passion) type bugs do not necessary cause any exceptions. Your code simply does not work without giving any indication why. Just provide a wrongly typed parameter to endpoint creation and schema synchronization waiting code does work without any warnings in python driver. Great language indeed. Cannot be triggered with single node test, so may be indeed we should do only those to have fun :)

Reply:
@nyh Thanks for the detailed arguments!

My "Python is evil." was strictly about my distaste for dynamically typed languages. I just find it much more efficient to write large systems in a statically typed language, and as you mentioned the set of tests and supporting infrastructure can get really large.

>>Here I bring you back to my philosophy that the amount of this "library" stuff should be minimized.

I can't agree with this statement, it is too broad. Some tests don't need much infrastructure, likely because the interface between the system under test and the test itself is narrow - having the system set up once you can exercise it many times, like the CQL endpoint in the single Scylla node cluster. If we have many nodes, we get more knobs to tune (gossiper, raft group0 state etc) and they all requires attention and thought. That said, I also agree that there is a big room for improvement in `test/pylib`, e.g. cluster reuse feature doesn't provide much value, most of the time each test allocates a fresh new cluster spending more time to set up than it could be.

I also believe that good infrastructure is paramount to making tests stable, usable, and easy to write. You've given several bullets why the current Scylla C++ tests are far from ideal, and I agree with them! But the problem is not the C++ language by itself. In most cases we can solve the problems by better infra, or by 'retargeting' the test to a more stable API. I think the integration/functional tests should strive to stick to the most stable API the system provides, and the user facing API is the ideal one - it can't be changed voluntary, so when we refactor the system the tests doesn't break. We can write the tests in C++ (or in rust, or in any statically typed language) using the the same public client driver API.

All in all, there two major differences in our approaches:
* I believe that "strong typing" brings a lot of value;
* I don't hesitate about investing in testing infrastructure. And in general, I tend to consider tests as ordinary program code, subject to the same general principles of code reuse and abstractions, which is essentially what is meant by infrastructure.

Reply:
> In my limited experience with the Python (which I hate with a passion) type bugs do not necessary cause any exceptions. Your code simply does not work without giving any indication why.

This is also my not-so-limited experience with Python. Before I joined Scylla I worked for a year on a large project written fully in Python, based on Django. And it wasn't unusual to have a production system run for a few days or weeks even, and then at some point the server throws internal error code 500... and in the logs you see "`int` object is not callable". So even though some indication is there, it's really not obvious why we tried to use `int` as a function after the system was running for few weeks, how did the `int` end up in this place, and generally it's a huge PITA.

Reply:
If folk feel so strongly about type checking, these days, at least partially,  they can add it to their Python code (https://github.com/python/mypy for example is one way I'm familiar with)

Reply:
> 
> I'm not saying that Python can't create bugs. I'm not saying that wrong types can't be a reason for these bugs. But I am claiming that for the types of use cases I've been using it (single-node cql-pytest and multi-node dtest) maybe 1% of my test development time was chasing type-related bugs. And I'm saying I had far worse experiences in writing Scylla C++ tests than I had in writing Python tests.

Of course Python (and other scripting language) has its use, and the use case you describe looks fitting indeed. But there are more complicated tests (cluster tests) that require a lot of code (to mange those clusters) and then it becomes a relatively large project with multiple people participating and at this point typeless languages are not a good fit (not that writing cluster management in C++ would be fun or even sensible).

Reply:
> I can't agree with this statement, it is too broad. Some tests don't need much infrastructure, likely because the interface between the system under test and the test itself is narrow - having the system set up once you can exercise it many times, like the CQL endpoint in the single Scylla node cluster. If we have many nodes, we get more knobs to tune (gossiper, raft group0 state etc) and they all requires attention and thought. That said, I also agree that there is a big room for improvement in `test/pylib`, e.g. cluster reuse feature doesn't provide much value, most of the time each test allocates a fresh new cluster spending more time to set up than it could be.

All of this is true, but what I would personally like to see (and this is orthogonal to the C++/Python issue) is that after a feature is added to the test infrastructure, then 20 tests would be written to make use of this feature. Adding 20 features and then writing two tests is, in my opinion, the wrong way to do thing, and creates the kind of frustration you're describing - you end up thinking that Python sucks for tests because you're spending most of your time writing a library, not writing tests, and perhaps Python does suck for writing a library.

> I also believe that good infrastructure is paramount to making tests stable, usable, and easy to write. You've given several bullets why the current Scylla C++ tests are far from ideal, and I agree with them! But the problem is not the C++ language by itself. In most cases we can solve the problems by better infra, or by 'retargeting' the test to a more stable API. I think the integration/functional tests should strive to stick to the most stable API the system provides, and the user facing API is the ideal one - it can't be changed voluntary, so when we refactor the system the tests doesn't break. We can write the tests in C++ (or in rust, or in any statically typed language) using the the same public client driver API.

I agree. This is, by the way, what the Cassandra guys did: They started by writing functional tests in Python - dtest, and writing unit tests which use internal APIs in Java. But then they realized how bad dtest is (and it wasn't because of Python!), so they moved all the single-node tests to Java - still keeping those tests to use the CQL request API, not internal functions (except for a few unfortunate exceptions). These are the tests I am in the process (about 50% done) to translate to Python for our framework.

We could have done this in C++. We didn't. I still think Python is better and makes it easier to write tests, but I agree that with some work we could have made it better in C++ as well. 

> All in all, there two major differences in our approaches:
> 
>     * I believe that "strong typing" brings a lot of value;
>     * I don't hesitate about investing in testing infrastructure. And in general, I tend to consider tests as ordinary program code, subject to the same general principles of code reuse and abstractions, which is essentially what is meant by infrastructure.

It's not that I disagree with those things (I don't), but I have an even more important goal as my primary goal: I want to make it easy to write and read tests. I want the test infrastructure to make it easier to write tests of the type we want to write. I don't care as much whether the test infrastructure makes it easier to add more and more features test infrastructure.

Maybe we can come to a middle ground: It's ok to use classes, strong typing, mypy, linters, etc., etc., on the test infrastructure - but don't require any of that on the tests themselves :-)

In any case, I disagree that Python is an evil language for writing tests. I actually think it's the best language for it. That's why I started test/alternator in Python and not in C++, Sh, Perl, Java, Pascal, Fortran, Tcl, or any other language I know :-)

Reply:
> Maybe we can come to a middle ground: It's ok to use classes, strong typing, mypy, linters, etc., etc., on the test infrastructure - but don't require any of that on the tests themselves :-)

I fully agree. I would also prefer to keep writing tests as easy and fun as possible. I strongly prefer writing tests in python, and one of the reason is the lack of all the boilerplate around typing in C++. It is so refreshing to not have to think ahead of what the type of each variable has to be, and then do all the clunky type conversions.

I think using types in python is very valuable in the framework code, it serves as extra documentation. It is notoriously hard to find out what a python function returns, sometimes you have to chase the return value across multiple functions (and then you miss the comma and don't realize it's actually a tuple).

Reply:
To summaries: nobody wants to specify  types, but want everybody else to do so.

Reply:
I don't actually see a significant difference between regular tests code and infrastructure code. Suppose you wrote some tests, you have some common logic between them, you extracted this logic in a function or class, is this already an infrastructure or not? Then someone saw this helpful function/class and reused it elsewhere. Then someone else wants to fix those places that are now using this function / class, but without an explicitly specified type, they will have to figure them out manually each time, based on how this function or class is implemented or used.

In theory you can manage this by rigorous code style/review process, e.g. you are now allowed to reuse something without first annotating it with the type hints. Actually, this is similar to how @kbr-scylla conducts his reviews now. But it's not fun - `mypy` is not automatically enforced and it's easy to forget to run it. And it's crucial - if you already have a working code it's difficult to motivate yourself to add these annotations if they are not enforced by the compiler in the first place. We basically have two languages combined in the same codebase. You write the working code in the language without annotations (just because it's easy) and then you need to convert to the other language manually, and the only motivation for this is making the life of your colleagues easy. You need a lot of dedication for this.

Reply:
> To summaries: nobody wants to specify types, but want everybody else to do so.

ChatGPT to the rescue?

Reply:
> I don't actually see a significant difference between regular tests code and infrastructure code. Suppose you wrote some tests, you have some common logic between them, you extracted this logic in a function or class, is this already an infrastructure or not?

It's not. I think that if you want it to be "infrastructure" you need to think long and hard and justify your pet convenience function becoming infrastructure. This is one of the things that most bothers me when I need to read or write dtests: They almost never use the CQL or DynamoDB library directly - they always do it through layers of classes and functions which do a lot of things together that some thought everyone will need. Then, when the next person came along, and the convenience function no longer did exactly what the new test did, that next person modifies the convenience function to add another parameter and do yet another thing. The result is very hard to understand or to write. Compare this to a test function which just makes CQL calls to write exactly what it needs to the table, and the reader can just... shudder... read the commands in sequence and know what the test does.

> Then someone saw this helpful function/class and reused it elsewhere. Then someone else wants to fix those places that are now using this function / class, but without an explicitly specified type, they will have to figure them out manually each time, based on how this function or class is implemented or used.

There's also a small thing called comments ;-)

> In theory you can manage this by rigorous code style/review process, e.g. you are not allowed to reuse something without first annotating it with the type hints. Actually, this is similar to how @kbr-scylla conducts his reviews now. But it's not fun - `mypy` is not automatically enforced and it's easy to forget to run it. And it's crucial - if you already have a working code it's difficult to motivate yourself to add these annotations if they are not enforced by the compiler in the first place.

You still didn't explain to me why it's "crucial". You are describing a process nobody wants to do and needs to be forced on developers, and in some rare cases saves you 5 minutes of misunderstanding. It doesn't sound "crucial" to me ;-)

> We basically have two languages combined in the same codebase. You write the working code in the language without annotations (just because it's easy) and then you need to convert to the other language manually, and the only motivation for this is making the life of your colleagues easy. You need a lot of dedication for this.

You are assuming that if we all just used type hints, it would make the life of the test developers easier. I just never saw any evidence for that. The evidence I have from personal experience is that the fewer requirements you make from the test developers, the fewer boilerplate stuff and magic incintations you need to just write a one-line test, and fewer draconic auto-checkers before you can commit it - the easier it becomes and the more developers are able and motivated to include good tests with their code.

By the way, this long thread started not complaining about mypy but about Python itself, which is an evil language to write tests in. I still think it's the other way around. But I also strongly believe that it's a good thing that people can write tests in many different frameworks - C++, single-node or multi-node tests in Python, textual ("approval tests"), dtest, SCT, etc. Each person can choose which framework he prefers, and the better framework will eventually get more tests written in it (in this sense, dtest is the best framework!). If someone prefers C++, they can work on improving the C++ test framework.

Reply:
> I don't actually see a significant difference between regular tests code and infrastructure code. Suppose you wrote some tests, you have some common logic between them, you extracted this logic in a function or class, is this already an infrastructure or not? Then someone saw this helpful function/class and reused it elsewhere. Then someone else wants to fix those places that are now using this function / class, but without an explicitly specified type, they will have to figure them out manually each time, based on how this function or class is implemented or used.

I think this is very rare in general. I never go hunting for code I could reuse in other tests. I only look in `conftest.py` for common fixtures. I may reuse code I wrote earlier, that I remember about. But only if the new test is in the same file.

I fully agree with the "don't repeat yourself principle", but I think forcing code reuse at all costs is just as bad. It leads to these overly complicated utility monstrosities that you see in dtests, that have to cater to 100 different use-cases and are not convenient for any of them. But people are forced to use them via code-review because copying code is bad.

This might be controversial but in recent years I actually grown to like copying code. It is liberating to just copy some code, and completely tailor it to your use case, throwing away all the things you don't need and fully integrating it into its current context. I find, this often leads to simpler and easier to understand code. And in some cases, it leads to *shorter* code too.

Reply:
FWIW https://github.com/scylladb/scylla_s3_reloc_server used type hints from day one and it was very helpful to prevent type issues.
We had one exception that was fixed in https://github.com/scylladb/scylla_s3_reloc_server/pull/66 where a member was assigned a value with mismatching type, and the member didn't have a type hint, unfortunately.

I don't think having the type hints in test code (in contrast to test libraries) is a big overhead, and if it can enable sanitation by mypy (that can run in CI for example, even if we don't run it locally in test.py), I think it's worth it.